File: rest/src/main/java/com/jgoticks/TicketSeller.java
package com.jgoticks;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.PoisonPill;
import akka.actor.Props;
import akka.event.DiagnosticLoggingAdapter;
import akka.event.Logging;
import akka.japi.Option;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.Value;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class TicketSeller extends AbstractActor {
    private final static String ACTOR_TICKET_SELLER = "ticketSeller";

    private String event_;
    private List<Ticket> tickets_ = new ArrayList<>();
    private final DiagnosticLoggingAdapter log = Logging.getLogger(this);

    public static Props props(String event) {
        return Props.create(TicketSeller.class, () -> new TicketSeller(event));
    }

    public static String name() {
        return ACTOR_TICKET_SELLER;
    }

    private TicketSeller(String event) {
        event_ = event;
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Add.class, msg ->
                    tickets_.addAll(msg.getTickets()))
                .match(Buy.class, msg -> {
                    if (tickets_.size() < msg.getTickets()) {
                        log.info("ticket count exceed. count=" + tickets_.size() + ", request=" + msg.getTickets());
                        getSender().tell(Tickets.of(event_, Collections.emptyList()), getSelf());
                    } else {
                        List<Ticket> entries = IntStream.range(0, msg.getTickets())
                                .mapToObj(i -> tickets_.remove(0))
                                .collect(Collectors.toList());
                        getSender().tell(Tickets.of(event_, entries), getSelf());
                    }
                })
                .match(GetEvent.class, msg ->
                    getSender().tell(Option.some(new BoxOffice.Event(event_, tickets_.size())), getSelf()))
                .match(Cancel.class, msg -> {
                    getSelf().tell(PoisonPill.getInstance(), ActorRef.noSender());
                    getSender().tell(Option.some(new BoxOffice.Event(event_, tickets_.size())), getSelf());
                })
                .matchAny(o -> log.info("received unknown message."))
                .build();
    }

    static @AllArgsConstructor @Data class Ticket {
        private final int id;
    }

    static  @Value(staticConstructor ="of") class Tickets {
        private String event;
        private List<Ticket> tickets;
    }

    static @Data class Add {
        private final List<Ticket> tickets;
    }

    static @Data class Buy {
        private final int tickets;
    }

    static class GetEvent {}
    static class Cancel {}
}

File: rest-remote/src/main/scala/com/goticks/TicketSeller.scala
package com.goticks

import akka.actor.{Actor, PoisonPill, Props}

object TicketSeller {
  def props(event: String) = Props(new TicketSeller(event))

  case class Ticket(id: Int)
  case class Tickets(event: String, tickets: Vector[Ticket] = Vector.empty[Ticket])
  case class Add(tickets: Vector[Ticket])
  case class Buy(tickets: Int)
  case object GetEvent
  case object Cancel
}

class TicketSeller(event: String) extends Actor {
  import TicketSeller._

  var tickets = Vector.empty[Ticket]

  override def receive() = {
    case Add(newTickets) => tickets = tickets ++ newTickets
    case Buy(numberOfTickets) =>
      val entries = tickets.take(numberOfTickets)
      if (entries.size >= numberOfTickets) {
        sender() ! Tickets(event, entries)
        tickets = tickets.drop(numberOfTickets)
      } else sender() ! Tickets(event)
    case GetEvent => sender() ! Some(BoxOffice.Event(event, tickets.size))
    case Cancel =>
      self ! PoisonPill
      sender() ! Some(BoxOffice.Event(event, tickets.size))
  }
}
